backend/backends/statsdaemon/statsdaemon.go

```diff
@@ -76,49 +78,125 @@ func (client *client) SendMetrics(ctx context.Context, metrics *types.MetricMap)
 
 	conn, err := net.Dial("udp", client.addr)
 	if err != nil {
-		return fmt.Errorf("error connecting to statsd backend: %s", err)
+		return fmt.Errorf("[%s] error connecting: %v", BackendName, err)
 	}
 	defer conn.Close()
 
-	var lastError error
-	buf := new(bytes.Buffer)
+	return processMetrics(metrics, func(buf *bytes.Buffer) (*bytes.Buffer, error) {
+		_, err := buf.WriteTo(conn)
+		if err != nil {
+			return nil, fmt.Errorf("[%s] error sending: %v", BackendName, err)
+		}
+		return buf, nil
+	})
+}
+
+// SendMetricsAsync flushes the metrics to the statsd server, preparing payload synchronously but doing the send asynchronously.
+func (client *client) SendMetricsAsync(ctx context.Context, metrics *types.MetricMap, c backendTypes.SendCallback) {
+	if metrics.NumStats == 0 {
+		c(nil)
+		return
+	}
+	localCtx, cancelFunc := context.WithCancel(ctx)
+
+	datagrams := make(chan *bytes.Buffer, sendChannelSize)
+
+	conn, err := net.Dial("udp", client.addr)
+	if err != nil {
+		c(fmt.Errorf("[%s] error connecting: %v", BackendName, err))
+		return
+	}
+	go func() {
+		defer conn.Close()
+		defer cancelFunc() // Tell the processMetrics function to stop if it is still running
+		for {
+			select {
+			case <-localCtx.Done():
+				c(localCtx.Err())
+				return
+			case buf, ok := <-datagrams:
+				if !ok {
+					c(nil)
+					return
+				}
+				_, errWrite := buf.WriteTo(conn)
+				buf.Reset() // In case of error we must reset the buffer before returning to the pool
+				bufFree.Put(buf)
+				if errWrite != nil {
+					c(fmt.Errorf("[%s] error sending: %v", BackendName, errWrite))
+					return
+				}
+			}
+		}
+	}()
+	err = processMetrics(metrics, func(buf *bytes.Buffer) (*bytes.Buffer, error) {
+		select {
+		case <-localCtx.Done(): // This can happen if 1) parent context is Done or 2) receiver encountered an error
+			return nil, localCtx.Err()
+		case datagrams <- buf:
+			return bufFree.Get().(*bytes.Buffer), nil
+		}
+	})
+	if err == nil {
+		// All metrics sent to the channel and context wasn't cancelled (yet) - consuming goroutine will exit
+		// with success (unless ctx gets a cancel after the close, which is also ok)
+		close(datagrams)
+	} else if err != context.Canceled && err != context.DeadlineExceeded {
+		log.Panicf("Unexpected error: %v", err)
+	}
+}
+
+func processMetrics(metrics *types.MetricMap, handler overflowHandler) (retErr error) {
+	type failure struct {
+		err error
+	}
+	defer func() {
+		if r := recover(); r != nil {
+			if f, ok := r.(failure); ok {
+				retErr = f.err
+			} else {
+				panic(r)
```

Comment:
Initially I wrote it using usual approach (return `error`, check with `if`) but it was ugly and I had to introduce another iteration method `EachWhile()` for counters, timers, gauges and sets. I decided to try to apply the approach with `panic`/`recover` described here https://blog.golang.org/defer-panic-and-recover and it looks MUCH better.

This particular line just re-panics a recovered panic in case it was something else that panicked, not our code (e.g. division by zero somewhere).